akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala

```diff
@@ -325,26 +325,37 @@ object HttpRequest {
    * include a valid [[akka.http.scaladsl.model.headers.Host]] header or if URI authority and [[akka.http.scaladsl.model.headers.Host]] header don't match.
    */
   def effectiveUri(uri: Uri, headers: immutable.Seq[HttpHeader], securedConnection: Boolean, defaultHostHeader: Host): Uri = {
-    def findHost(headers: immutable.Seq[HttpHeader]): OptionVal[Host] = {
+    def findHost: OptionVal[Host] = {
       val it = headers.iterator
       while (it.hasNext) it.next() match {
         case h: Host ⇒ return OptionVal.Some(h)
         case _       ⇒ // continue ...
       }
       OptionVal.None
     }
-    val hostHeader: OptionVal[Host] = findHost(headers)
+    val hostHeader = findHost
     if (uri.isRelative) {
       def fail(detail: String) =
         throw IllegalUriException(
           s"Cannot establish effective URI of request to `$uri`, request has a relative URI and $detail; " +
             "consider setting `akka.http.server.default-host-header`")
-      val Host(host, port) = hostHeader match {
+      val Host(hostHeaderHost, hostHeaderPort) = hostHeader match {
         case OptionVal.None                 ⇒ if (defaultHostHeader.isEmpty) fail("is missing a `Host` header") else defaultHostHeader
         case OptionVal.Some(x) if x.isEmpty ⇒ if (defaultHostHeader.isEmpty) fail("an empty `Host` header") else defaultHostHeader
         case OptionVal.Some(x)              ⇒ x
       }
-      uri.toEffectiveHttpRequestUri(host, port, securedConnection)
+      def isWebsocket: Boolean = {
```

Comment:
Could we somehow prevent the extra scan through the header list? I'm not sure how critical it is but would be nice to prevent that every request has to go through it. (In the middle-term, we should probably try to evolve the `HttpRequest` model to include those kind of flags as I started to discuss in #991.)